arwen/ethapi/ethereumei.go

```diff
@@ -619,18 +619,18 @@ func ethrevert(context unsafe.Pointer, dataOffset int32, length int32) {
 	metering := arwen.GetMeteringContext(context)
 
 	data, err := runtime.MemLoad(dataOffset, length)
-	if withFault(err, context) {
+	if arwen.WithFault(err, context, false) {
```

Comment:
ethereum API error handling is still wrong. Right now, as the code is, any execution of the ethereum API consumes all the gasLeft, so there is nothing the developer can do anymore even if the code goes back to the wasmer instance. For those methods where there is no return value, the execution can be stopped. 
For methods which have return 0 / 1 / -1 only the API gas cost has to be deducted, not all the remaining.
In case of revert, there is no need to make any processing. Just use all the gas and runTime.SignalUserError revert.

Reply:
Still wrong, indeed. Fixed as follows:
* `arwen.WithFault()` now only consumes remaining gas if `failExecution == true`
* Ethereum API methods with *no return value* now call `WithFault` with `failExecution == true` 
* Ethereum API methods with *integer return value* now call `WithFault` with `failExecution == false`, leaving the gas unconsumed
* Changed Ethereum API methods to always consume gas first, before any possible failure

Reply:
Revert seems correct to me now as well.

Reply:
Extra question here, @camilbancioiu, @sasurobert.
According to the [docs](https://ewasm.readthedocs.io/en/mkdocs/eth_interface/), some functions return 2 on revert:

```
result: i32 - Returns 0 on success, 1 on failure and 2 on revert
```

But we do not have any `return 2` in our `ethereumei.go` - are we missing something?

Reply:
we have to return as those docs say

Reply:
Seems like need a full conformity check on our Ethereum API, but that's beyond this PR. Created a JIRA issue for it: https://elrondnetwork.atlassian.net/browse/EN-5772